core/trino-main/src/main/java/io/trino/execution/scheduler/PipelinedStageExecution.java

```diff
@@ -224,23 +231,35 @@ public synchronized void schedulingComplete()
             return;
         }
 
-        if (isFlushing()) {
-            stateMachine.transitionToFlushing();
+        try {
+            allTasksLock.readLock().lock();
+            if (isFlushing(flushingOrFinishedTasksCounter.get())) {
+                stateMachine.transitionToFlushing();
+            }
+            if (isFinished(finishedTasksCounter.get())) {
+                stateMachine.transitionToFinished();
+            }
         }
-        if (finishedTasks.containsAll(allTasks)) {
-            stateMachine.transitionToFinished();
+        finally {
+            allTasksLock.readLock().unlock();
         }
 
         for (PlanNodeId partitionedSource : stage.getFragment().getPartitionedSources()) {
             schedulingComplete(partitionedSource);
         }
     }
 
-    private synchronized boolean isFlushing()
+    private boolean isFlushing(long flushingOrFinishedTasks)
     {
         // to transition to flushing, there must be at least one flushing task, and all others must be flushing or finished.
-        return !flushingTasks.isEmpty()
-                && allTasks.stream().allMatch(taskId -> finishedTasks.contains(taskId) || flushingTasks.contains(taskId));
+        // allTasks and allTasksCounter is protected by allTasksLock so that number of total tasks was not changed
+        return flushingOrFinishedTasks == allTasksCounter && wasFlushingTask;
+    }
+
+    private boolean isFinished(long finishedTasks)
+    {
+        // allTasks and allTasksCounter is protected by allTasksLock so that number of total tasks was not changed
+        return finishedTasks == allTasksCounter;
```

Comment:
Why can't you use just `allTasks.size()`. If you are not allowing concurrent modifications to `allTasks` use of size is fine (you could not modify it concurrently anyway as this is a normal HashMap)

Reply:
I can do it. I use the other counter have the `long`. If you call `allTasks.size()` you get `int`. Where the size of the set is greater then `INT_MAX` then `size()` returns `INT_MAX`. If you think that `INT_MAX` is sufficient I think that we should to it - it make code simpler and less complex.